File: AnimalSaver/Assets/Standard Assets (Mobile)/Scripts/Cow.js
#pragma strict
private var Cow:OTSprite;  
private var Cow_die = false;
private var landWater:OTSprite;
private var landBoat:OTSprite;  
private var landBoatPosX:float;  
private var startPos:Vector3;
private var k:float;
function Start () 
{
	landWater = null;
	landBoat = null;
	landBoatPosX = 0f;
	// Get this cow's sprite
	Cow = GetComponent("OTSprite");
	Cow.InitCallBacks(this);
	Cow.frameIndex = 1;
	Cow.rigidbody.isKinematic = false;
	// initialize variables
	// initialize the velocity as 25 in -y direction
	Cow.rigidbody.velocity = new Vector3(0, -25,0);
 	
 	var randomValue1:float = Random.value;
 	// here 3*Cow.size.y is to give the gap on left and right. Please increase it to decrease difficulty, 
 	// because it will minimize the falling range on the sky 
 	// start position x should be random
 	var creationScreenwidth = Screen.width - 3*Cow.size.y;
	startPos.x =  creationScreenwidth*(randomValue1 - 0.5);
	startPos.y = Screen.height/2;
	startPos.z = 0;
	Cow.position = new Vector2(startPos.x , startPos.y);

	// give random initilial direction (left or right)
	var randomValue2:float = Random.value;
	if(randomValue2 < 0.5)
		k = startPos.y*0.015-(3.1415926/2);
	else
		k = startPos.y*0.015+(3.1415926/2);
}

function Update () 
{
	var targetPos: Vector3 = transform.position;
	var tmpY:float =  targetPos.y;
	var tmpX:float =  targetPos.x;
	

	// if it lands on one boat, let it move with boat
	if(landBoat != null) {
		tmpX += landBoat.position.x - landBoatPosX;
		landBoatPosX = landBoat.position.x;
		Cow.position = new Vector2(tmpX,tmpY);
		if (Cow.outOfView)
		{
		 // Destroy the object
		 Destroy(Cow.gameObject);
		}
		return;
	} else if (landWater != null) { 
		 // if it lands in water, let it die
		 Cow.frameIndex = 0;
		 // set the die flag
		 Cow_die = true;
		 Cow.depth = 0.5; // on back
		 if (Cow.outOfView)
		{
		 // Destroy the object
		 Destroy(Cow.gameObject);
		}  
		return;
	}
	
	// respond to the left arrow
	if(Input.GetKey(KeyCode.LeftArrow) && Cow_die == false)
	{
		Cow.frameIndex = 2;
		//tmpX -= 30 * Time.deltaTime;
		startPos.x+=-30 * Time.deltaTime;

	}
	// respond to the right arrow
	else if(Input.GetKey(KeyCode.RightArrow) && Cow_die == false)
	{
		Cow.frameIndex = 3;
		//tmpX += 30 * Time.deltaTime;
		startPos.x += 30 * Time.deltaTime;
	}
	// default down picture
	else if (Cow_die == false)
	{
		Cow.frameIndex = 1;
	}
	// Reset the position according to interaction
	if (Cow_die == false){
		tmpX = startPos.x+(1/((0.012)+0.00005*Time.time))* Mathf.Cos(tmpY*0.015-k);
		Cow.position = new Vector2(tmpX,tmpY);
	}
	
	// determine the border of the object's movement
	// the middle x position is: 
	if(Cow.position.x  < - (Screen.width - Cow.size.x)/2)
	{
		// show the Cow dead object 
		Cow.frameIndex = 0;
		// set the die flag
		Cow_die = true;
	}
	else if(Cow.position.x >  (Screen.width- Cow.size.x)/2)
	{
		// show the Cow dead right object 
		Cow.frameIndex = 4;
		Cow.depth += 0.5; // on back 
		// set the die flag
		Cow_die = true;
	} 
	if (Cow.outOfView)
	{
	 	// Destroy the object
	 	Destroy(Cow.gameObject);
	 	// Test Game Over function
		Application.LoadLevel(2);

	}
}
public function OnStay(owner:OTObject)
{
	var obj:OTObject = owner.collisionObject;  
	//var box:BoxCollider= null;
	//	box = GetComponent(BoxCollider);
	if(Cow_die == false
     	&& (obj.name == "Boat0" || obj.name == "Boat1") 
 		&& (obj.position.x+obj.size.x/2) >= (Cow.position.x+Cow.size.x/2)
 		&& (obj.position.x-obj.size.x/2) <= (Cow.position.x-Cow.size.x/2)
 		&& (obj.position.y+obj.size.y/2) >= (Cow.position.y-Cow.size.y/6))
    {
    	landBoat = owner.collisionObject;
    	landBoatPosX = landBoat.position.x;
    	Cow.depth = 0.5; // on back 
    	Cow.rigidbody.velocity.y = 0;
    	Cow.collidable = false; // no need enter onStay any more
    } 
//    else if(collide with Water){
//    	// show the Cow dead left object 
//		Cow.frameIndex = 0;
//		//Cow.depth += 0.5; // on back 
//		// set the die flag
//		Cow_die = true;
//		Cow.depth = 0.5; // on back  
//		Cow.collidable = false; // no need enter onStay any more
//    }
}

//function OnCollision(owner:OTObject)
//{
//	// a collision occured
//	OT.print(owner.name+" collided with "+owner.collisionObject.name+" at "+owner.collision.contacts[0].point);
//}

//function onOutOfView(owner:OTObject)
//{
//    OT.DestroyObject(owner);
//}

File: AnimalSaver/Assets/Standard Assets (Mobile)/Scripts/Cow.cs
using UnityEngine;
using System.Collections;

[RequireComponent(typeof(AudioSource))]
public class Cow : MonoBehaviour {
	
	OTSprite cow;  
	bool Cow_die = false;
	//OTObject landWater;
	OTObject landBoat;  
	float landBoatPosX;  
	Vector3 startPos;
	float k;
	Vector3 mouseDownPosition;
	bool bWindAffected;
	float tWind;
	float deadTime;
	float dragstep;
	AudioSource sound;
	AudioSource died;
	//AudioSource generated;
	// Use this for initialization
	void Start () {
		//landWater = null;
		landBoat = null;
		landBoatPosX = 0f;
		
		// Get this chick's sprite
		cow = GetComponent<OTSprite>();
		cow.InitCallBacks(this);
		cow.onInput = OnInput;	
		cow.onEnter = OnCollided;
		cow.frameIndex = 1;
		// initialize variables
		// initialize the velocity and start position
	 	cow.rigidbody.isKinematic = false;
		// initialize the velocity as 25 in -y direction
		cow.rigidbody.velocity = new Vector3(0, -50,0);
		cow.size = new Vector2(Screen.width *0.255f, Screen.width *0.255f*1.333f);
		
	 	float randomValue1 = Random.value;
	 	// here 3*Cow.size.y is to give the gap on left and right. Please increase it to decrease difficulty, 
	 	// because it will minimize the falling range on the sky 
	 	// start position x should be random
		var creationScreenwidth = Screen.width - 3*cow.size.y;
		startPos.x =  creationScreenwidth*(randomValue1 - 0.5f);
		startPos.y = Screen.height/2;
		startPos.z = 0;
		cow.position = new Vector2(startPos.x , startPos.y);
		
		// give random initilial direction (left or right)
		float randomValue2 = Random.value;
		if(randomValue2 < 0.5f)
			k = startPos.y*0.015f-(3.1415926f/2f);
		else
			k = startPos.y*0.015f+(3.1415926f/2f);
		
		bWindAffected = false;
		tWind = 0;
		deadTime = 0f;
		dragstep = Screen.height * 0.05f;
		
		AudioSource[] aSources = GetComponents<AudioSource>();
    	sound = aSources[0];//Drag
    	died = aSources[1];//dead
		

	}
	
	// Update is called once per frame
	void Update () {
		Vector3 targetPos = transform.position;
		float tmpY =  targetPos.y;
		float tmpX =  targetPos.x;
		//tmpY = startPos.y +  cow.rigidbody.velocity.y * Time.time;
		
		// if it lands on one boat, let it move with boat
		if(landBoat != null) {
			tmpX += landBoat.position.x - landBoatPosX;
			landBoatPosX = landBoat.position.x;
			cow.position = new Vector2(tmpX,tmpY);
			if (cow.outOfView)
			{
			 // Destroy the object
			 Destroy(cow.gameObject);
			 Main.decreaseAnimal(0);
			}
			return;
		}

		if(Cow_die == false)
		{
			if (bWindAffected)
			{
				//OT.print(Main.WindDirection);
				// respond to the left arrow
				if(Main.WindDirection == "left")
				{
					cow.frameIndex = 2;
					startPos.x += -dragstep * Time.deltaTime;
					sound.Play();
				}
				else if(Main.WindDirection == "right")
	
				{
					cow.frameIndex = 3;
					startPos.x += dragstep * Time.deltaTime;
					sound.Play();			
				}
				else if(Main.WindDirection == "down")
				{
					cow.frameIndex = 1;
					//startPos.y -= 30 * Time.deltaTime;
					tmpY -=  dragstep * Time.deltaTime;
					startPos.y -=  dragstep * Time.deltaTime;
					sound.Play();
				}
				else
					cow.frameIndex = 1;
				// Reset the position according to interaction	
				
				if (Main.WindDirection == "left" || 
					Main.WindDirection == "right" ||
					Main.WindDirection == "down")
					tWind += Time.deltaTime;
			}
		}
		else
		{
			cow.frameIndex = 0;
			Main.gameOver = true;
		}
	
			
		//if (tWind > Main.windDuration || Main.WindDirection == "none")
		if (bWindAffected && tWind > Main.windDuration)
		{
			bWindAffected = false;
			tWind = 0;
		}
		
		// Reset the position according to interaction
		if (Cow_die == false){
			tmpX = startPos.x+(1f/((0.012f)+0.00005f*Time.time))* Mathf.Cos(tmpY*0.015f-k);
			if (tmpX  < - (Screen.width - cow.size.x)/2)
				tmpX = -(Screen.width - cow.size.x)/2;
			if (tmpX >  (Screen.width- cow.size.x)/2)
				tmpX = (Screen.width- cow.size.x)/2;
			//tmpY = startPos.y -  cow.rigidbody.velocity.y * Time.time;
			cow.position = new Vector2(tmpX,tmpY);

		} // end if (cow_die == false)
		else //Cow_die == true
		{
			cow.rigidbody.velocity = new Vector3(0, cow.rigidbody.velocity.y  - 9.8f*(Time.time - deadTime), 0);
		}

		if (cow.outOfView)
		{
		 	// Destroy the object
		 	Destroy(cow.gameObject);
			Main.decreaseAnimal(0);
		 	// Game Over if any animal died
			if(Cow_die == true)
				Application.LoadLevel(2);
	
		}
	}

	public void OnCollided(OTObject owner)
	{
		OTObject obj = owner.collisionObject; 
		if(Cow_die == false
	     	&& (obj.name == "Boat0" || obj.name == "Boat1")) 
	    {
			// make sure the animal is ON the boat
			if((cow.position.x-cow.size.x/2)< (obj.position.x-obj.size.x/2))
				cow.position = new Vector2(obj.position.x-obj.size.x/2+ cow.size.x/2,cow.position.y);
			else if((cow.position.x+cow.size.x/2)> (obj.position.x+obj.size.x/2))//-0.25*obj.size.x is a tolerance
				cow.position = new Vector2(obj.position.x+obj.size.x/2- cow.size.x/2-0.25f*obj.size.x,cow.position.y);
	    	landBoat = owner.collisionObject;
	    	landBoatPosX = landBoat.position.x;
	    	cow.frameIndex = 1;
	    	cow.rigidbody.velocity = new Vector3(0, 0, 0);
	    	cow.collidable = false; // no need enter OnCollided any more
			
			Main.animalSaved++;
	    } else if(Cow_die == false && landBoat == null
				&& (obj.name.StartsWith("BackGround_Bottom")  ||  obj.name.StartsWith("Bee_left")
			||obj.name.StartsWith("Bee_right") ||obj.name.StartsWith("Maneater_plant")))
			{
				died.Play();
				deadTime = Time.time;
		    	// show the Cow dead left object 
				cow.frameIndex = 0;
				// set the die flag
				Cow_die = true;
				cow.collidable = false; // no need enter onStay any more
		    }

	}
	//Move to event to main window so that mouse does not need click on object percisely
	//This one is local detection. set register input to true in designer
	void OnInput(OTObject owner)
	{
		//OT.print("OnInput " + owner.name);
		bWindAffected = true;
	}
	
}
